README.md
# gvSIG-bpd Business Process Definition

Overview

  Is the business definition and publication Web tool for collaborative communities, using standard concepts from the Object Management Group. Built in Python as a Zope/Plone product. Depends on Plone product ModelDDvlPlone, hosted at OSOR at https://joinup.ec.europa.eu/page/osor.eu/projects/mddvlplone with forge at http://forge.osor.eu/projects/mddvlplone Runs on Zope2.9/Plone2.5. Tested on Debian Ubuntu 8 and Windows (TM) XP and Vista. Developed under contract by the company Model Driven Development sociedad limitada, in Valencia, Spain, and Antonio Carrasco Valero, for the local government ministry Conselleria de Infraestructuras y Transporte de la Generalidad Valenciana. 
  
  
Dependencies

  - "ModelDDvlPlone", itself dependent on "Archetypes", "Relations", "PloneLanguageTool"
  
  Installation will fail if these products are not 
  correctly deployed and available in the Plone instance.
  
  The installer will automatically install the prerequisite
  products, if corretly deployed and available in the Plone instance.
  
  
Copyright

  Copyright (c) 2008 by 2008 Model Driven Development sl and Antonio Carrasco Valero
  

Project Home

  - "ModelDD.org":http://www.ModelDD.org 

  
  
Authors

  - "Antonio Carrasco Valero and "Model Driven Development sl  Valencia (Spain)":http://www.ModelDD.org 

  
License

  GNU General Public License (GPL)
 
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.
 
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
  02110-1301, USA.


gvSIGbpd/release_notes.txt
gvSIG Business Process Definition RELEASE NOTES
===============================================

|
|

version 1.0.0.7
---------------


sk Id  	Task Summary  	Start Date  	End Date  	Percent Complete
606	BUG: Fails to paste Plone content elements (i.e. ATImage instances)	2009-12-07 17:00	  2009-12-14 17:00	100%
579	PERF04 Avoid high number of log entries when (re)installing the gvSIG-bpd product, which uses Relations, from Zope catalogs complaining about entries not found for missing relation instances	2009-11-24 17:00	  2009-12-01 17:00	100%
586	RELS02 Add relationship Send step delivers to zero or many Reception step	2009-11-26 00:00	  2009-12-03 00:00	100%


|
|

--------------------------------------------------------

|
|


**As by v200911211405**

|
|


Summary of use cases implemented after delivery of version 1.0 (build200906110445)
==================================================================================


- Add concept of Extension Point in the sequence of Business Process Steps, and

- the concept of Business Process extension

- Add the concept of Feature of Artefact

- Add the concepts of Using an Artefact or a Feature from a Business Process Step

- Add the relations: a Tool instructions are Business Processes, and a Business Process applies Tools.

- Search for words in all string fields, not just the fields named "Title", "Description" and "Text"

- Reorder aggregated elements by Drag&Drop.



**Use Cases already reported as implemented after delivery of version 1.0 (build200907160507**





- The application now automatically creates in the Plone Site an instance of ModelDDvlPloneTool,
  if one does not exist yet.
  
- Export an element as .zip archives with information in a character encoded (utf8) .xml file, 
  and plone binary content as separate files.

- Import into an element as .zip archives with .xml file and binary content.

- Cut/Copy/Paste/Delete on elements, or groups of elements.

- When pasting a Business Rule into a collection of Business Policies,
  refactor the Business Rule into a Business Policy,
  and vice-versa. Same for Reception/Input, Sends/Outputs,
  Tool/Artefact, Organizational Unit/Profile.
  
- Create a new Version as a copy of a whole root organization and its contents, 
  maintaining traceability links from elements in the new version 
  to elements in the previous version.  
  
- Standard Plone actions are intercepted or aliased to maintain a tight contron on application instances.
  
  
|
|


------------------------------

|
|



version 1.0.0.6
---------------

|
|


**RESOLVED**

Error Exception when creating a new version of a root organization

Error when Refactoring a Business Process Input as an Output

ETTY 04x01 Relationship &quot;Extends&quot; between Business Processes

ETTY 04x02 Add to Business Processes features to specify the management of Artefacts

ETTY 04x03 Add Text field to Business Process  different from the Purpose field

ETTY 04x04 Add to Business Processes features to specify the management of Artefact Features

ETTY 33 Extension Point Step

ETTY 34 Feature

ETTY10x01 Input Add features to specify minimum and maximum multiplicity or artefacts

ETTY11x01 Output Add features to specify minimum and maximum multiplicity or artefacts

ETTY13x1 Add to General Step features to specify the usage  production and consumption of Artefacts

ETTY13x2 Add to General Step features to specify the usage  production and consumption of Features in Artefacts

ETTY20x01 Reception Add features to specify minimum and maximum multiplicity or artefacts

ETTY21x01 SendAdd features to specify minimum and maximum multiplicity or artefacts

IMPO01x1 Import elements from XML file not in a zip archive with binary content

Import a whole root organization fails to create relations

SUC04 Search elements including text and string fields other than description and text

USEX 06  In Textual View  hide titles of previous Process Steps

USEX 07 In Textual View for an Artifact  display the title of the related Inputs\' Business Process owner

USEX 08 Add object action to reorder aggregated elements by drag and drop


|
|

**KNOWN BUGS Pending to be resolved in this iteration**

-None at this time-

|
|



**POSTPONED**


*To 1.1*

COMU 02 Edit and publish pages to present the gvSIG-bpd product to OSOR communities

DOCU 01 Edit and publish gvSIG-bpd user documentation

DOCU 02 Edit and publish gvSIG-bpd install documentation

DOCU 03 Edit and publish gvSIG-bpd administration documentation

DOCU04 Edit and publish gvSIG-bpd testing documentation and supporting artefacts

|
|

*Until further scheduling*


gvSIGbpd/BPDEnvio.py
# -*- coding: utf-8 -*-
#
# File: BPDEnvio.py
#
# Copyright (c) 2011 by Conselleria de Infraestructuras y Transporte de la
# Generalidad Valenciana
#
# GNU General Public License (GPL)
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.
#
#

__author__ = """Conselleria de Infraestructuras y Transporte de la Generalidad Valenciana
<gvSIGbpd@gvSIG.org>, Model Driven Development sl <gvSIGbpd@ModelDD.org>,
Antonio Carrasco Valero <carrasco@ModelDD.org>"""
__docformat__ = 'plaintext'

from AccessControl import ClassSecurityInfo
from Products.Archetypes.atapi import *
from Products.gvSIGbpd.BPDCondicional import BPDCondicional
from Products.gvSIGbpd.BPDConDatosDePrueba import BPDConDatosDePrueba
from Products.gvSIGbpd.BPDPasoGeneral import BPDPasoGeneral
from Products.gvSIGbpd.BPDConResolucionesDatos import BPDConResolucionesDatos
from Products.Relations.field import RelationField
from Products.gvSIGbpd.config import *

# additional imports from tagged value 'import'
from Products.ATReferenceBrowserWidget.ATReferenceBrowserWidget import ReferenceBrowserWidget

##code-section module-header #fill in your manual code here
##/code-section module-header

schema = Schema((

    IntegerField(
        name='multiplicidadMinima',
        widget=IntegerField._properties['widget'](
            label="Multiplicidad Minima",
            label2="Minimum Multiplicity",
            description="Numero minimo de elementos a enviar. Introduzca 0 para indicar que el Envio procedera con exito, aunque no se envie ningun artefacto.",
            description2="Minimum number of elements to send. Enter 0 to indicate that the Send will proceed with success, even if no Artefact of the type was sent",
            label_msgid='gvSIGbpd_BPDEnvio_attr_multiplicidadMinima_label',
            description_msgid='gvSIGbpd_BPDEnvio_attr_multiplicidadMinima_help',
            i18n_domain='gvSIGbpd',
        ),
        description="Numero minimo de elementos a enviar. Introduzca 0 para indicar que el Envio procedera con exito, aunque no se envie ningun artefacto.",
        duplicates="0",
        label2="Minimum Multiplicity",
        ea_localid="410",
        derived="0",
        precision=0,
        collection="false",
        styleex="volatile=0;",
        description2="Minimum number of elements to send. Enter 0 to indicate that the Send will proceed with success, even if no Artefact of the type was sent",
        ea_guid="{5CDF47ED-7EA6-4b5a-96E9-8535F338B5C8}",
        write_permission='Modify portal content',
        scale="0",
        default="0",
        label="Multiplicidad Minima",
        length="0",
        containment="Not Specified",
        position="3",
        owner_class_name="BPDEnvio"
    ),

    IntegerField(
        name='multiplicidadMaxima',
        widget=IntegerField._properties['widget'](
            label="Multiplicidad Maxima",
            label2="Maximum Multiplicity",
            description="Numero maximo de elementos a enviar. Introduzca -1 para indicar que no hay limite superior para el numero de Artefactos a enviar.",
            description2="Maximum number of elements to send. Enter -1 to indicate that there is no upper limit in the number of elements to send.",
            label_msgid='gvSIGbpd_BPDEnvio_attr_multiplicidadMaxima_label',
            description_msgid='gvSIGbpd_BPDEnvio_attr_multiplicidadMaxima_help',
            i18n_domain='gvSIGbpd',
        ),
        description="Numero maximo de elementos a enviar. Introduzca -1 para indicar que no hay limite superior para el numero de Artefactos a enviar.",
        duplicates="0",
        label2="Maximum Multiplicity",
        ea_localid="407",
        derived="0",
        precision=0,
        collection="false",
        styleex="volatile=0;",
        description2="Maximum number of elements to send. Enter -1 to indicate that there is no upper limit in the number of elements to send.",
        ea_guid="{878AAA95-0490-4fad-932B-24C2F6CA3C83}",
        write_permission='Modify portal content',
        scale="0",
        default="-1",
        label="Multiplicidad Maxima",
        length="0",
        containment="Not Specified",
        position="4",
        owner_class_name="BPDEnvio"
    ),

    RelationField(
        name='destinatarios',
        inverse_relation_label="Destinatario de Envios",
        additional_columns=['abreviatura'],
        inverse_relation_description="Envios que se destinan a este Perfil o Unidad Organizacional.",
        description="Perfiles o Unidades Organizacionales a los que se destina el Envio.",
        relationship='BPDDestinatarios',
        inverse_relation_field_name='destinatarioDeEnvios',
        inverse_relation_label2="Receiver of Send steps",
        label2="Receivers",
        inverse_relation_description2="Send process steps addressed to this participant Profile or Organisational Unit.",
        widget=ReferenceBrowserWidget(
            label="Destinatarios",
            label2="Receivers",
            description="Perfiles o Unidades Organizacionales a los que se destina el Envio.",
            description2="Participant Profiles or Organisational Units to whom this is Sent",
            label_msgid='gvSIGbpd_BPDEnvio_rel_destinatarios_label',
            description_msgid='gvSIGbpd_BPDEnvio_rel_destinatarios_help',
            i18n_domain='gvSIGbpd',
        ),
        label="Destinatarios",
        description2="Participant Profiles or Organisational Units to whom this is Sent",
        multiValued=1,
        containment="Unspecified",
        inverse_relationship='BPDDestinatarioDeEnvios',
        owner_class_name="BPDEnvio",
        deststyle="Union=0;Derived=0;AllowDuplicates=0;Owned=0;Navigable=Unspecified;"
    ),

    ComputedField(
        name='detallesPaso',
        widget=ComputedField._properties['widget'](
            label="Detalles del Paso",
            label2="Step details",
            description="Detalles acerca de las caracteristicas del Paso de Proceso de Negocio.",
            description2="Details about the fetarues of the Busienss Process Step",
            label_msgid='gvSIGbpd_BPDEnvio_attr_detallesPaso_label',
            description_msgid='gvSIGbpd_BPDEnvio_attr_detallesPaso_help',
            i18n_domain='gvSIGbpd',
        ),
        exclude_from_values_paragraph="True",
        description="Detalles acerca de las caracteristicas del Paso de Proceso de Negocio.",
        duplicates="0",
        label2="Step details",
        ea_localid="278",
        derived="0",
        precision=0,
        collection="false",
        styleex="volatile=0;",
        description2="Details about the fetarues of the Busienss Process Step",
        ea_guid="{5273BF3F-BBC0-4973-8DA6-2D1863879182}",
        exclude_from_values_form="True",
        scale="0",
        label="Detalles del Paso",
        length="0",
        containment="Not Specified",
        position="2",
        owner_class_name="BPDEnvio",
        expression="context.fTFLVsUnless([ ['esInicial', False,], ['titulosDestinatarios','',],['titulosArtefactosEnviados','',],['ejecutores',None,],[ 'titulosArtefactosUsados','',],[ 'titulosCaracteristicasUsadas','',],])",
        computed_types="string"
    ),

    RelationField(
        name='recepciones',
        inverse_relation_label="Envios",
        additional_columns=['tituloProcesoDeNegocio', 'titulosDestinatarios', 'titulosArtefactosRecibidos'],
        inverse_relation_description="Pasos de Envio llevados a cabo por otros roles, que entregan informacion a este paso de recepcion en este rol.",
        description="Pasos de Recepcion llevados a cabo por otros roles, a los que entrega informacion este paso de envio desde este rol.",
        relationship='BPDEnviosRecepciones',
        inverse_relation_field_name='envios',
        inverse_relation_label2="Sends",
        label2="Receptions",
        inverse_relation_description2="Send steps carried out by other roles, delivering information to this reception step by this role",
        widget=ReferenceBrowserWidget(
            label="Recepciones",
            label2="Receptions",
            description="Pasos de Recepcion llevados a cabo por otros roles, a los que entrega informacion este paso de envio desde este rol.",
            description2="Reception steps carried out by other roles, to which this send step delivers information from this role.",
            label_msgid='gvSIGbpd_BPDEnvio_rel_recepciones_label',
            description_msgid='gvSIGbpd_BPDEnvio_rel_recepciones_help',
            i18n_domain='gvSIGbpd',
        ),
        label="Recepciones",
        description2="Reception steps carried out by other roles, to which this send step delivers information from this role.",
        multiValued=1,
        containment="Unspecified",
        inverse_relationship='BPDRecepcionesEnvios',
        owner_class_name="BPDEnvio",
        deststyle="Union=0;Derived=0;AllowDuplicates=0;Owned=0;Navigable=Unspecified;"
    ),

    RelationField(
        name='artefactosEnviados',
        inverse_relation_label="Enviado desde Pasos de Procesos de Negocio",
        additional_columns=['codigo'],
        inverse_relation_description="Pasos de Proceso de Negocio que envian un Artefacto de este tipo a un Participante externo.",
        description="Artefactos que se envian en este paso a un Participante externo.",
        relationship='BPDArtefactosEnviados',
        inverse_relation_field_name='pasosQueEnvianElArtefacto',
        inverse_relation_label2="Sent from Business Process Steps",
        label2="Sent Artefacts",
        inverse_relation_description2="Business Process Steps sending Artefacts of this type to external participants.",
        widget=ReferenceBrowserWidget(
            label="Artefactos Enviados",
            label2="Sent Artefacts",
            description="Artefactos que se envian en este paso a un Participante externo.",
            description2="Artefacts thar shall be sent while in this step to an external participant.",
            label_msgid='gvSIGbpd_BPDEnvio_rel_artefactosEnviados_label',
            description_msgid='gvSIGbpd_BPDEnvio_rel_artefactosEnviados_help',
            i18n_domain='gvSIGbpd',
        ),
        label="Artefactos Enviados",
        description2="Artefacts thar shall be sent while in this step to an external participant.",
        multiValued=1,
        containment="Unspecified",
        inverse_relationship='BPDPasosQueEnvianElArtefacto',
        owner_class_name="BPDEnvio",
        deststyle="Navigable=Unspecified;Union=0;Derived=0;AllowDuplicates=0;Owned=0;"
    ),

    ComputedField(
        name='titulosArtefactosEnviados',
        widget=ComputedField._properties['widget'](
            label="Artefactos Enviados",
            label2="Sent Artefacts",
            description="Artefactos que se envian en este paso a un Participante externo.",
            description2="Artefacts thar shall be sent while in this step to an external participant.",
            label_msgid='gvSIGbpd_BPDEnvio_attr_titulosArtefactosEnviados_label',
            description_msgid='gvSIGbpd_BPDEnvio_attr_titulosArtefactosEnviados_help',
            i18n_domain='gvSIGbpd',
        ),
        exclude_from_values_paragraph="True",
        description="Artefactos que se envian en este paso a un Participante externo.",
        duplicates="0",
        label2="Sent Artefacts",
        ea_localid="270",
        derived="0",
        collection="false",
        styleex="volatile=0;",
        description2="Artefacts thar shall be sent while in this step to an external participant.",
        containment="Not Specified",
        ea_guid="{FEE95966-1F45-469b-8A13-A9746A0E65E0}",
        position="1",
        owner_class_name="BPDEnvio",
        label="Artefactos Enviados",
        expression="context.fFV('artefactosEnviados')",
        exclude_from_values_form="True"
    ),

    ComputedField(
        name='titulosDestinatarios',
        widget=ComputedField._properties['widget'](
            label="Destinatarios",
            label2="Receivers",
            description="Perfiles o Unidades Organizacionales a los que se destina el Envio.",
            description2="Participant Profiles or Organisational Units to whom this is Sent",
            label_msgid='gvSIGbpd_BPDEnvio_attr_titulosDestinatarios_label',
            description_msgid='gvSIGbpd_BPDEnvio_attr_titulosDestinatarios_help',
            i18n_domain='gvSIGbpd',
        ),
        exclude_from_values_paragraph="True",
        description="Perfiles o Unidades Organizacionales a los que se destina el Envio.",
        duplicates="0",
        label2="Receivers",
        ea_localid="271",
        derived="0",
        collection="false",
        label="Destinatarios",
        description2="Participant Profiles or Organisational Units to whom this is Sent",
        containment="Not Specified",
        ea_guid="{AF512C2B-BF61-4a84-8E8A-4E954298FEE2}",
        position="0",
        owner_class_name="BPDEnvio",
        styleex="volatile=0;",
        expression="context.fFV('destinatarios')",
        exclude_from_values_form="True"
    ),

),
)

##code-section after-local-schema #fill in your manual code here
##/code-section after-local-schema

BPDEnvio_schema = OrderedBaseFolderSchema.copy() + \
    getattr(BPDCondicional, 'schema', Schema(())).copy() + \
    getattr(BPDConDatosDePrueba, 'schema', Schema(())).copy() + \
    getattr(BPDPasoGeneral, 'schema', Schema(())).copy() + \
    getattr(BPDConResolucionesDatos, 'schema', Schema(())).copy() + \
    schema.copy()

##code-section after-schema #fill in your manual code here
##/code-section after-schema

class BPDEnvio(OrderedBaseFolder, BPDCondicional, BPDConDatosDePrueba, BPDPasoGeneral, BPDConResolucionesDatos):
    """
    """
    security = ClassSecurityInfo()
    __implements__ = (getattr(OrderedBaseFolder,'__implements__',()),) + (getattr(BPDCondicional,'__implements__',()),) + (getattr(BPDConDatosDePrueba,'__implements__',()),) + (getattr(BPDPasoGeneral,'__implements__',()),) + (getattr(BPDConResolucionesDatos,'__implements__',()),)

    # This name appears in the 'add' box
    archetype_name = 'Envio'

    meta_type = 'BPDEnvio'
    portal_type = 'BPDEnvio'


    # Change Audit fields

    creation_date_field = 'fechaCreacion'
    creation_user_field = 'usuarioCreador'
    modification_date_field = 'fechaModificacion'
    modification_user_field = 'usuarioModificador'
    deletion_date_field = 'fechaEliminacion'
    deletion_user_field = 'usuarioEliminador'
    is_inactive_field = 'estaInactivo'
    change_counter_field = 'contadorCambios'
    sources_counters_field = 'contadoresDeFuentes'
    change_log_field = 'registroDeCambios'




    # Versioning and Translation fields

    inter_version_field = 'uidInterVersionesInterno'
    version_field = 'versionInterna'
    version_storage_field = 'versionInternaAlmacenada'
    version_comment_field = 'comentarioVersionInterna'
    version_comment_storage_field = 'comentarioVersionInternaAlmacenada'
    inter_translation_field = 'uidInterTraduccionesInterno'
    language_field = 'codigoIdiomaInterno'
    fields_pending_translation_field = 'camposPendientesTraduccionInterna'
    fields_pending_revision_field = 'camposPendientesRevisionInterna'



    allowed_content_types = [] + list(getattr(BPDCondicional, 'allowed_content_types', [])) + list(getattr(BPDConDatosDePrueba, 'allowed_content_types', [])) + list(getattr(BPDPasoGeneral, 'allowed_content_types', [])) + list(getattr(BPDConResolucionesDatos, 'allowed_content_types', []))
    filter_content_types             = 1
    global_allow                     = 0
    content_icon = 'bpdenvio.gif'
    immediate_view                   = 'Textual'
    default_view                     = 'Textual'
    suppl_views                      = ('Textual', 'Tabular', )
    typeDescription                  = "Representa un Paso de Proceso en que se envia una informacion al exterior del ambito del Proceso."
    typeDescMsgId                    =  'gvSIGbpd_BPDEnvio_help'
    archetype_name2                  = 'Send'
    typeDescription2                 = '''Represents a Business Process Step where some information is sent outside of the boudaries of the Business Process.'''
    archetype_name_msgid             = 'gvSIGbpd_BPDEnvio_label'
    factory_methods                  = None
    factory_enablers                 = None
    propagate_delete_impact_to       = None


    actions =  (


       {'action': "string:$object_url/content_status_history",
        'category': "object",
        'id': 'content_status_history',
        'name': 'State',
        'permissions': ("View",),
        'condition': """python:0"""
       },


       {'action': "string:${object_url}/MDDInspectClipboard",
        'category': "object_buttons",
        'id': 'inspectclipboard',
        'name': 'Clipboard',
        'permissions': ("View",),
        'condition': """python:object.fAllowRead()"""
       },


       {'action': "string:${object_url}/Editar",
        'category': "object",
        'id': 'edit',
        'name': 'Edit',
        'permissions': ("Modify portal content",),
        'condition': """python:object.fAllowWrite()"""
       },


       {'action': "string:${object_url}/MDDOrdenar",
        'category': "object_buttons",
        'id': 'reorder',
        'name': 'Reorder',
        'permissions': ("Modify portal content",),
        'condition': """python:object.fAllowWrite()"""
       },


       {'action': "string:${object_url}/MDDExport",
        'category': "object_buttons",
        'id': 'mddexport',
        'name': 'Export',
        'permissions': ("View",),
        'condition': """python:object.fAllowExport()"""
       },


       {'action': "string:${object_url}/MDDImport",
        'category': "object_buttons",
        'id': 'mddimport',
        'name': 'Import',
        'permissions': ("Modify portal content",),
        'condition': """python:object.fAllowImport()"""
       },


       {'action': "string:${object_url}/sharing",
        'category': "object",
        'id': 'local_roles',
        'name': 'Sharing',
        'permissions': ("Manage properties",),
        'condition': """python:1"""
       },


       {'action': "string:${object_url}/",
        'category': "object",
        'id': 'view',
        'name': 'View',
        'permissions': ("View",),
        'condition': """python:1"""
       },


       {'action': "string:${object_url}/MDDChanges",
        'category': "object_buttons",
        'id': 'mddchanges',
        'name': 'Changes',
        'permissions': ("View",),
        'condition': """python:1"""
       },


       {'action': "string:${object_url}/MDDVersions",
        'category': "object_buttons",
        'id': 'mddversions',
        'name': 'Versions',
        'permissions': ("View",),
        'condition': """python:1"""
       },


       {'action': "string:${object_url}/MDDCacheStatus/",
        'category': "object_buttons",
        'id': 'mddcachestatus',
        'name': 'Cache',
        'permissions': ("View",),
        'condition': """python:1"""
       },


       {'action': "string:${object_url}/TextualRest",
        'category': "object_buttons",
        'id': 'textual_rest',
        'name': 'TextualRest',
        'permissions': ("View",),
        'condition': """python:1"""
       },


    )

    _at_rename_after_creation = True

    schema = BPDEnvio_schema

    ##code-section class-header #fill in your manual code here
    ##/code-section class-header

    # Methods

    security.declarePublic('manage_afterAdd')
    def manage_afterAdd(self,item,container):
        """
        """
        
        return self.pHandle_manage_afterAdd(  item, container)

    security.declarePublic('manage_pasteObjects')
    def manage_pasteObjects(self,cb_copy_data=None,REQUEST=None):
        """
        """
        
        return self.pHandle_manage_pasteObjects( cb_copy_data, REQUEST)

    security.declarePublic('moveObjectsByDelta')
    def moveObjectsByDelta(self,ids,delta,subset_ids=None):
        """
        """
        
        return self.pHandle_moveObjectsByDelta( ids, delta, subset_ids=subset_ids)

registerType(BPDEnvio, PROJECTNAME)
# end of class BPDEnvio

##code-section module-footer #fill in your manual code here
##/code-section module-footer



